bitswap/network/connecteventmanager.go

```diff
@@ -32,7 +32,38 @@ type connectEventManager struct {
 
 type peerState struct {
 	newState, curState state
-	pending            bool
+	accepted           chan struct{}
+}
+
+func newPeerState() *peerState {
+	return &peerState{accepted: make(chan struct{})}
+}
+
+func (p *peerState) isPending() bool {
+	select {
+	case <-p.accepted:
+		return false
+	default:
+	}
+	return true
+}
+
+func (p *peerState) accept() {
+	select {
+	case <-p.accepted:
+	default:
+		close(p.accepted)
+	}
+}
+
+func (p *peerState) setPending() {
+	if !p.isPending() {
+		p.accepted = make(chan struct{})
```

Comment:
I'm a little unsure about this, is this possibly introducing another race where we replace the closed channel with a new one, unsynchronized. Is this safe to do when we're only reading from or closing the channel elsewhere? I'm not 100% sure about the atomicity of these operations.

Reply:
We rely on `connectEventManager`'s mutex to do most things, but there is a call to `waitAccept` that is unsynchronized, it just does a `<-p.accepted`; that could be happening while we're doing this operation of setting `accepted` to a new channel—although we're only doing this if the channel is already closed so `<-p.accepted` _should_ have been a quick noop; is it atomic?